### PR TITLE
fix(material/input): Do not set `aria-invalid` on empty `matInput`s

### LINKED ISSUE #22777: bug(matInput): Screen Readers do not announce that an empty `matInput[required]` is invalid
#### Reproduction
StackBlitz: https://components-issue-pcy3uq.stackblitz.io

Steps to reproduce:
1. Turn on a screen reader
2. Focus the input
3. Move focus away from the input
4. Return focus to the input

#### Expected Behavior
Screen reader announces to the user that the form field is invalid.

#### Actual Behavior
Screen reader does not announce to the user that the form field is invalid.

Note that while the value in `mat-error` is announced to the user when the field is focused there is no indication that the field itself is invalid (unless the error message itself contains that).

#### Environment
- Angular:^12.0.1
- CDK/Material: ^12.0.1
- Browser(s): Any
- Operating System (e.g. Windows, macOS, Ubuntu): Any

#### Details
This issue is rooted in the fact that `matInput` sets `aria-invalid="false"` for empty inputs.  This causes screen readers not to announce that a `required` input is invalid despite being empty.

We should update `matInput` to not apply `aria-invalid` on instances that do not have any value within the input.  This may be accomplished via changing [this line within the matInput directive](https://github.com/angular/components/blob/master/src/material/input/input.ts#L86) from:
```
'[attr.aria-invalid]': 'errorState && !empty'
```
to:
```
'[attr.aria-invalid]': 'empty ? null : errorState'
```

Doing this will allow us to comply with [the ARIA spec for `aria-invalid`](https://www.w3.org/TR/wai-aria/#aria-invalid) (since `aria-invalid` isn't applied to the element until the user has entered some data) while also providing the correct announcements (since `aria-required`/`input[required]` is indicating to the screen reader that the empty field is invalid).

##### Testing methodology

I used this HTML document for testing ARIA attribute behavior with VoiceOver on Chrome and JAWS on Chrome:
```
<html>
  <body>
    <form>
      <input required/>
      <button type="submit">Submit</button>
    </form>
  </body>
</html>
```
Testing with adding/removing various aria attributes on the `input` element showed the following behavior:

| ARIA attributes | value | JAWS+Chrome | VoiceOver+Chrome |
|------------|-------|-------------|------------------|
| `aria-required` not present, `aria-invalid` not present| no value | "Edit Required invalid entry" | "required, invalid data, edit text"|
| `aria-required` not present, `aria-invalid` not present| "value" | "Edit Required, value" | "value, required, edit text"|
| `aria-required="true"`, `aria-invalid` not present| no value | "Edit Required, invalid entry" | "required, invalid data, edit text" |
| `aria-required="true"`, `aria-invalid` not present | "value" | "Edit required, value" | "value, required, edit text" |
| `aria-required="true"`, `aria-invalid="true"` | no value | "Edit required, invalid entry" | "required, invalid data, edit text" |
| `aria-required="true"`, `aria-invalid="true"` | "value" | "Edit required, invalid entry" | "value, required, invalid data, edit text" |
| `aria-required="true"`, `aria-invalid="false"` | no value | "Edit required, type in text" | "required, edit text" |
| `aria-required="true"`, `aria-invalid="false"` | "value" | "Edit required, type in text" | "value, required, edit text" |

While I did not test with NVDA/ChromeVox or on FF/IE/etc. I'm reasonably confident that these results will replicate across these use cases as well.

### PR DESCRIPTION
**Pull request by ByzantineFailure**
Updates the logic for setting `aria-invalid` on `matInput` to not set
the value at all if the input has no value.

Prior to this PR `matInput` sets `aria-invalid="false"` for any empty
`matInput`, including `required` ones.  This suppresses screen readers'
announcement to users that such inputs are in an invalid state.

Fixes #22777

### CODE REVIEW

**Review by mmalerba — APPROVED** [review]
LGTM

**Comment by mmalerba** [comment]
It looks like there's a lint failure because we need to add the same logic/test for the MDC-based input

File: src/material/input/input.ts [hunk 1]
```diff
@@ -83,7 +83,7 @@ const _MatInputBase = mixinErrorState(class {
     '[attr.readonly]': 'readonly && !_isNativeSelect || null',
     // Only mark the input as invalid for assistive technology if it has a value since the
     // state usually overlaps with `aria-required` when the input is empty and can be redundant.
-    '[attr.aria-invalid]': 'errorState && !empty',
+    '[attr.aria-invalid]': 'empty ? null : errorState',
```
Comment (zelliott) [inline src/material/input/input.ts:86]: While we're here, is it _slightly_ more correct to change this to `'(empty && required) ? null : errorState'`? I'm thinking of an edge case where an input is empty, but not required, but still in an error state (maybe because there's validation that looks for duplicates and there's already an empty input). If so, we'd want to update the comment above as well.

Comment (crisbeto) [reply src/material/input/input.ts:86]: This behavior is intentional, see https://github.com/angular/components/pull/21609. There's also a linked issue.

**Review by crisbeto — CHANGES_REQUESTED** [review]

**Review by mmalerba — COMMENTED** [review]
I thought I remembered seeing some recent changes to this part of the code. It sounds like we may not want to make this change, but I'll leave it open for discussion for now

File: src/material/input/input.ts (hunk 1 above)
Comment (ByzantineFailure) [reply src/material/input/input.ts:86]: Yep!  The reason #21609 was requested was because without it `aria-invalid` was being applied to `aria-required` inputs with no value.  This violates the ARIA spec (see here : https://www.w3.org/TR/wai-aria/#aria-invalid ) and caused #18140 to be filed.

Unfortunately, setting `aria-invalid="false"` on these inputs suppresses any native indication to a screen reader user that the input is in an invalid state.  The only information they receive is the error message itself.  If worded correctly the error message can be enough (e.g. if it includes "Invalid" in the text) but this requires component users to be aware of this fact and actively account for it.

Without a well-worded error message, this results in a different experience for SR users than visual users -- the input is flagged as invalid and displayed prominently to visual users as invalid, but screen reader users get no native notification at all when they focus the input.

I think @zelliott might have a good point here that we may wish to do `(empty && required)` rather than simply `empty` for our `null` state.

File: src/material-experimental/mdc-input/input.spec.ts [hunk 2]
```diff
@@ -1081,7 +1081,7 @@ describe('MatMdcInput with forms', () => {
       expect(describedBy).toBe(errorIds);
     }));
 
-    it('should not set `aria-invalid` to true if the input is empty', fakeAsync(() => {
+    it('should not set `aria-invalid` if the input is empty', fakeAsync(() => {
```
Comment (ByzantineFailure) [inline src/material-experimental/mdc-input/input.spec.ts:1084]: Looking at this test in more detail.

Comment (ByzantineFailure) [reply src/material-experimental/mdc-input/input.spec.ts:1084]: Edge case:  This test tests an input where `Validators.required` is added to the form field but the actual `input` with the `matInput` directive does not have the `required` attribute.

In this case, the `aria-invalid` attribute should be `true` since `aria-required` is _not_ present on the element.  This is the exact edge case @zelliott was trying to catch!

**Comment by ByzantineFailure** [comment]
All right, it has been a minute since I've properly used git and I accidentally closed this branch when rebasing it to squash it down to one commit.  I'll... be right with you.

**Comment by ByzantineFailure** [comment]
I'mma just close and re-open this with the correctly-named branch.  Apologies, folks.
